### PR TITLE
#50 patch before payment

### PR DESCRIPTION
**Pull request by lojzatran**
Part of https://github.com/commercetools/commercetools-paypal-plus-integration/issues/50

This PR exposes new endpoint `/{tenantName}/commercetools/patch/payments/{ctpPaymentId}`. Invoice payments causes problems, user's address needs to be submitted before the user gets redirected to the payment page.

### CODE REVIEW

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 1]
```diff
@@ -154,21 +154,48 @@ public PaymentHandleResponse createPayment(@Nonnull String ctpPaymentId) {
...
+                    builder.add(patchBillingAddress);
+                }
+                AddInterfaceInteraction addInterfaceInteractionAction = createAddInterfaceInteractionAction(patchShippingAddress, REQUEST);
                 CompletionStage<PaymentHandleResponse> patchCS = ctpFacade.getPaymentService()
                         // 1. add patch request to the ctp payment interface interaction
                         .updatePayment(paymentId, Collections.singletonList(addInterfaceInteractionAction))
                         .thenCompose(payment -> paypalPlusFacade.getPaymentService()
                                 // 2. patch payment on paypal
-                                .patch(paypalPlusPayment, replace)
+                                .patch(paypalPlusPayment, builder.build())
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: I'd not recommend to use the `ImmutableBuilder` here, it doesn't give any benefits in this case (immutability, shrunk capacity etc). 

Having only 2 potential entries in the list it is enough to use any simple default Java list implementation (e.g. `ArrayList`)

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 2]
```diff
@@ -196,16 +223,8 @@ public PaymentHandleResponse executePayment(@Nonnull String paypalPlusPaymentId,
...
-                        PaymentHandleResponse paymentHandleResponse = patchAddress(cart, paypalPlusPaymentId);
-                        boolean isSuccessful = HttpStatus.valueOf(paymentHandleResponse.getStatusCode()).is2xxSuccessful();
-                        if (isSuccessful) {
-                            // execute payment only when patching was successful
-                            return updatePayerIdInCtpPayment(paypalPlusPaymentId, paypalPlusPayerId)
-                                    .thenCompose(ctpPayment -> executePaymentAndCreateTxn(paypalPlusPaymentId, paypalPlusPayerId, ctpPayment));
-                        } else {
-                            return CompletableFuture.completedFuture(paymentHandleResponse);
-                        }
+                        return updatePayerIdInCtpPayment(paypalPlusPaymentId, paypalPlusPayerId)
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: the payment in `updatePayerIdInCtpPayment()` not found by id, the value in the stage will be null, but chained `executePaymentAndCreateTxn()` accepts _ctpPayment_ only as _@Nonnull_ => NPE prone solution

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 3]
```diff
@@ -266,27 +285,50 @@ private AddInterfaceInteraction createAddInterfaceInteractionAction(@Nonnull Pay
...
                                         // 4. create charge transaction in the ctp payment
-                                        .thenApply(updatedCtpPayment -> createChargeTransaction(paypalPlusPaymentId, paypalPayment, updatedCtpPayment));
+                                        .thenApply(updatedCtpPayment2 -> createChargeTransaction(paypalPlusPaymentId, paypalPayment, updatedCtpPayment2));
                             });
                 })
                 .thenApply(ignore -> PaymentHandleResponse.ofHttpStatus(HttpStatus.CREATED));
     }
 
+    private CompletionStage<io.sphere.sdk.payments.Payment> updateCtpPayment(Payment paypalPayment, io.sphere.sdk.payments.Payment ctpPayment) {
+        ImmutableList.Builder<UpdateAction<io.sphere.sdk.payments.Payment>> builder = ImmutableList.builder();
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: Same here: list builder is overhead.

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 4]
```diff
@@ -266,27 +285,50 @@ private AddInterfaceInteraction createAddInterfaceInteractionAction(@Nonnull Pay
...
+        builder.add(createAddInterfaceInteractionAction(paypalPayment, RESPONSE));
+        PaymentInstruction paymentInstruction = paypalPayment.getPaymentInstruction();
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: Why we are sticking with deprecated objects?

> PaymentInstruction object is not available for public use.

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 5]
```diff
@@ -266,27 +285,50 @@ private AddInterfaceInteraction createAddInterfaceInteractionAction(@Nonnull Pay
...
+        listBuilder.add(SetCustomField.ofObject(ACCOUNT_HOLDER_NAME, paymentInstruction.getRecipientBankingInstruction().getAccountHolderName()));
+        listBuilder.add(SetCustomField.ofObject(IBAN, paymentInstruction.getRecipientBankingInstruction().getInternationalBankAccountNumber()));
+        listBuilder.add(SetCustomField.ofObject(BIC, paymentInstruction.getRecipientBankingInstruction().getBankIdentifierCode()));
+        listBuilder.add(SetCustomField.ofObject(PAYMENT_DUE_DATE, paymentInstruction.getPaymentDueDate()));
+        listBuilder.add(SetCustomField.ofObject(AMOUNT, Money.of(new BigDecimal(paymentInstruction.getAmount().getValue()),
+                paymentInstruction.getAmount().getCurrency())));
+        return listBuilder.build();
+    }
+
     private CompletionStage<io.sphere.sdk.payments.Payment> createChargeTransaction(@Nonnull String paypalPlusPaymentId,
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: pls, put the overloaded methods (`createChargeTransaction`) next to each other in the class file.

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 4 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/paypal/PayPal-Java-SDK/issues/313

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 6]
```diff
@@ -154,21 +154,48 @@ public PaymentHandleResponse createPayment(@Nonnull String ctpPaymentId) {
...
             Optional<String> paymentIdOpt = getCtpPaymentId(cartWithPaymentsExpansion, paypalPlusPaymentId);
             return paymentIdOpt.map(paymentId -> {
                 Payment paypalPlusPayment = new Payment().setId(paypalPlusPaymentId);
-                ShippingAddress shippingAddress = shippingAddressMapper.ctpAddressToPaypalPlusAddress(cartWithPaymentsExpansion.getShippingAddress());
-                Patch replace = new Patch("add", "/transactions/0/item_list/shipping_address").setValue(shippingAddress);
-                AddInterfaceInteraction addInterfaceInteractionAction = createAddInterfaceInteractionAction(replace, REQUEST);
+                ImmutableList.Builder<Patch> builder = ImmutableList.builder();
+
+                io.sphere.sdk.models.Address shippingAddress = cartWithPaymentsExpansion.getShippingAddress();
+                Patch patchShippingAddress = new Patch("add", "/transactions/0/item_list/shipping_address").setValue(addressMapper.ctpAddressToPaypalPlusShippingAddress(shippingAddress));
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: If it's null, then it's an unexpected error anyway, but I will check and throw some more informative exception.

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 1 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/5288d7243d7e90fc0bfc661e5fd710dfa327b84e

**Comment by lojzatran** [comment]
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/5288d7243d7e90fc0bfc661e5fd710dfa327b84e

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 3 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/5288d7243d7e90fc0bfc661e5fd710dfa327b84e

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/ffbe68ba90314226042c65c300d7150ae90ef186

[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/cf77c936c946fcd6b9085012b4afb8c9123fc70b

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 5 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/e838e31a00f58885521f23850ff2cfbf07a26d8c

Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/6cb6c76fa37871969a1aae7f2c2937e13c7265be

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 2 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/9673a67ac2ff6b29b3ee4e35a124b8f051f1ce0f

Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: You misunderstood me: i complain that next chained function `executePaymentAndCreateTxn()` will fail if payment in `updatePayerIdInCtpPayment()` not found

**Comment by andrii-kovalenko-ct** [comment]
only https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51#discussion_r136617984 left

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java (hunk 2 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/51/commits/74e58720b5606484cc91819dfa6dc5f55174221d

**Pull request merged**